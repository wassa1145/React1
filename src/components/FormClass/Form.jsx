import React, { Component } from 'react';
import { Button } from './Button';
import { Textarea } from './Textarea';
import { Message } from './Message';

export class Form extends Component {
	state = {
		name: 'Отправить',
		value: '',
		message: 'Привет! Я дефолтное сообщение. Можешь меня заменить',
	}
	handleClick = () => {
		this.setState({ message: this.state.value })
		this.setState({ value: '' })
	}
	handleChange = (event) => {
		this.setState({ value: event.target.value })
	}
  render() {
		return <div className='form'>
			<Message message={this.state.message}/>
			<Textarea change={this.handleChange} value={this.state.value}/>
			<Button name={this.state.name} click={this.handleClick} disabled={this.state.value ? false : true}/>
		</div> 
  }
}
