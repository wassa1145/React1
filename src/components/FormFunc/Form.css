.form {
  width: 100%;
  padding: 20px;
  margin: 0 auto;
  box-sizing: border-box;
	margin-top: 30px;
  display: flex;
  justify-content: space-between;
}
