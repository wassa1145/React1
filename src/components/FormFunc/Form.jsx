import React, { useState, useCallback } from 'react';
import { Button } from './components/Button/Button';
import { Textarea } from './components/Textarea/Textarea';
import './Form.css';

export const Form = ({ addMessage }) => {
  const [message, setMessage] = useState('');

  const handleClick = (event) => {
    event.preventDefault();
    addMessage(message);
    setMessage('');
  };
  const changeMessage = useCallback((event) => {
    setMessage(event.target.value);
  }, []);

  return (
    <form className="form" onSubmit={handleClick} data-testid="form">
      <Textarea
        change={changeMessage}
        message={message}
        placeholder="Введите сообщение"
      />
      <Button name="Отправить" disabled={!message} />
    </form>
  );
};
