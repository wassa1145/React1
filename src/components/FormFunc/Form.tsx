import React, { useState, useEffect, FC, memo, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { ThunkDispatch } from 'redux-thunk';
import { ChatsState } from 'src/store/chats/reducer';
import { AddMessage } from 'src/store/chats/types';
import { CONSTANTS } from 'src/constants';
import { addMessageWithReply } from 'src/store/chats/slice';
import { Button } from 'src/components/FormFunc/components/Button/Button';
import { Textarea } from 'src/components/FormFunc/components/Textarea/Textarea';
import './Form.css';

export const Form: FC = memo(() => {
  const [message, setMessage] = useState('');
  const textFieldRef = useRef<HTMLInputElement>(null);
  const { chatId } = useParams();
  const dispatch =
    useDispatch<ThunkDispatch<ChatsState, void, ReturnType<AddMessage>>>();

  const handleClick = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (chatId && message) {
      dispatch(
        addMessageWithReply({
          chatId,
          message: { author: CONSTANTS.USER, value: message },
        })
      );
    }
    setMessage('');
    textFieldRef.current?.focus();
  };

  useEffect(() => {
    textFieldRef.current?.focus();
  }, []);

  return (
    <form className="form" onSubmit={handleClick} data-testid="form">
      <Textarea
        change={(event) => setMessage(event.target.value)}
        message={message}
        textFieldRef={textFieldRef}
        placeholder="Введите сообщение"
      />
      <Button name="Отправить" disabled={!message} />
    </form>
  );
});
