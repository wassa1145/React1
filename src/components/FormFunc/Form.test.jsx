import React from 'react';
import { Form } from './Form';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

describe('Form', () => {
  render(<Form />);
  it('render component', () => {
    expect(screen.getByTestId('form')).toBeInTheDocument();
  });

  it('render with snapshot', () => {
    const { asFragment } = render(<Form />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('render textarea and button', () => {
    render(<Form />);
    expect(screen.getByTestId('textarea')).toBeInTheDocument();
    expect(screen.getByTestId('submit-button')).toBeInTheDocument();
  });

  it('clear textarea on subbmit', async () => {
    const mockHandler = jest.fn();
    render(<Form addMessage={mockHandler} />);
    await userEvent.type(screen.getByTestId('textarea'), 'new message');
    await userEvent.click(screen.getByTestId('submit-button'));
    expect(screen.getByTestId('textarea')).toHaveValue('');
  });
});
