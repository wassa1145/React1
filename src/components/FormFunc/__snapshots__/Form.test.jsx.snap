// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render with snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
    data-testid="form"
  >
    <textarea
      class="form__textarea"
      data-testid="textarea"
      placeholder="Введите сообщение"
      rows="6"
    />
    <button
      class="form__button"
      data-testid="submit-button"
      disabled=""
      type="submit"
    >
      Отправить
    </button>
  </form>
</DocumentFragment>
`;
